// _viewport-mercator-project@5.0.0-alpha.2@viewport-mercator-project/dist/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _webMercatorViewport = require('./web-mercator-viewport');

Object.defineProperty(exports, 'default', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_webMercatorViewport).default;
  }
});
Object.defineProperty(exports, 'WebMercatorViewport', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_webMercatorViewport).default;
  }
});
Object.defineProperty(exports, 'PerspectiveMercatorViewport', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_webMercatorViewport).default;
  }
});
Object.defineProperty(exports, 'fitBounds', {
  enumerable: true,
  get: function get() {
    return _webMercatorViewport.fitBounds;
  }
});

var _webMercatorUtils = require('./web-mercator-utils');

Object.defineProperty(exports, 'projectFlat', {
  enumerable: true,
  get: function get() {
    return _webMercatorUtils.projectFlat;
  }
});
Object.defineProperty(exports, 'unprojectFlat', {
  enumerable: true,
  get: function get() {
    return _webMercatorUtils.unprojectFlat;
  }
});
Object.defineProperty(exports, 'getMeterZoom', {
  enumerable: true,
  get: function get() {
    return _webMercatorUtils.getMeterZoom;
  }
});
Object.defineProperty(exports, 'getDistanceScales', {
  enumerable: true,
  get: function get() {
    return _webMercatorUtils.getDistanceScales;
  }
});
Object.defineProperty(exports, 'getWorldPosition', {
  enumerable: true,
  get: function get() {
    return _webMercatorUtils.getWorldPosition;
  }
});
Object.defineProperty(exports, 'getViewMatrix', {
  enumerable: true,
  get: function get() {
    return _webMercatorUtils.getViewMatrix;
  }
});
Object.defineProperty(exports, 'getUncenteredViewMatrix', {
  enumerable: true,
  get: function get() {
    return _webMercatorUtils.getUncenteredViewMatrix;
  }
});
Object.defineProperty(exports, 'getProjectionMatrix', {
  enumerable: true,
  get: function get() {
    return _webMercatorUtils.getProjectionMatrix;
  }
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
