// _deck.gl@5.0.1@deck.gl/dist/react/experimental/viewport-layout.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _flatten = require('../../core/utils/flatten');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var ViewportLayout = function (_React$Component) {
  _inherits(ViewportLayout, _React$Component);

  function ViewportLayout(props) {
    _classCallCheck(this, ViewportLayout);

    return _possibleConstructorReturn(this, (ViewportLayout.__proto__ || Object.getPrototypeOf(ViewportLayout)).call(this, props));
  }

  // Iterate over viewport descriptors and render viewports at specified positions


  _createClass(ViewportLayout, [{
    key: '_renderChildrenUnderViewports',
    value: function _renderChildrenUnderViewports() {
      // Flatten out nested viewports array
      var viewports = (0, _flatten.flatten)(this.props.viewports, { filter: Boolean });

      // Build a viewport id to viewport index
      var viewportMap = {};
      viewports.forEach(function (viewportOrDescriptor) {
        var viewport = viewportOrDescriptor.viewport || viewportOrDescriptor;
        if (viewport.id) {
          viewportMap[viewport.id] = viewport;
        }
      });

      var originalChildren = _react2.default.Children.toArray(this.props.children);
      var lastElement = originalChildren.pop();

      var children = originalChildren.map(function (child, i) {
        // If viewportId is provided, match with viewport
        if (child.props.viewportId) {
          var viewport = viewportMap[child.props.viewportId];

          // TODO - this is too react-map-gl specific
          var newProps = Object.assign({}, child.props, {
            visible: viewport.isMapSynched(),
            width: viewport.width,
            height: viewport.height
          }, viewport.getMercatorParams());

          var clone = (0, _react.cloneElement)(child, newProps);

          var _style = {
            position: 'absolute',
            left: viewport.x,
            top: viewport.y,
            width: viewport.width,
            height: viewport.height
          };

          child = (0, _react.createElement)('div', { key: 'viewport-component-' + i, style: _style }, clone);
        }

        return child;
      });

      var style = { position: 'absolute', left: 0, top: 0 };
      children.push((0, _react.createElement)('div', { key: 'children', style: style }, lastElement));

      return children;
    }
  }, {
    key: 'render',
    value: function render() {
      // Render the background elements (typically react-map-gl instances)
      // using the viewport descriptors
      var children = this._renderChildrenUnderViewports();
      return (0, _react.createElement)('div', {}, children);
    }
  }]);

  return ViewportLayout;
}(_react2.default.Component);

exports.default = ViewportLayout;
