// _deck.gl@5.0.1@deck.gl/dist/experimental-layers/src/text-layer/font-atlas.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.makeFontAtlas = makeFontAtlas;

var _luma = require('luma.gl');

var MAX_CANVAS_WIDTH = 1024; /* global document */

var fontSize = 64;
var padding = 2;

var charList = [];
for (var i = 32; i < 128; i++) {
  charList.push(String.fromCharCode(i));
}

function setTextStyle(ctx, fontFamily) {
  ctx.font = fontSize + 'px ' + fontFamily;
  ctx.fillStyle = '#000';
  ctx.textBaseline = 'hanging';
  ctx.textAlign = 'left';
}

function makeFontAtlas(gl, fontFamily) {
  var canvas = document.createElement('canvas');
  var ctx = canvas.getContext('2d');
  setTextStyle(ctx, fontFamily);

  // measure texts
  var row = 0;
  var x = 0;
  var mapping = {};

  charList.forEach(function (char) {
    var _ctx$measureText = ctx.measureText(char),
        width = _ctx$measureText.width;

    if (x + width > MAX_CANVAS_WIDTH) {
      x = 0;
      row++;
    }
    mapping[char] = {
      x: x,
      y: row * (fontSize + padding),
      width: width,
      height: fontSize,
      mask: true
    };
    x += width;
  });

  canvas.width = MAX_CANVAS_WIDTH;
  canvas.height = (row + 1) * (fontSize + padding);

  setTextStyle(ctx, fontFamily);
  for (var char in mapping) {
    ctx.fillText(char, mapping[char].x, mapping[char].y);
  }

  return {
    mapping: mapping,
    texture: new _luma.Texture2D(gl, {
      pixels: canvas
      // no need to specify texture filter here
      // icon layer use the most accurate mode by default
    })
  };
}
