// _deck.gl@5.0.1@deck.gl/dist/core/shaderlib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.lighting = exports.project64 = exports.project = exports.picking = exports.fp64 = exports.fp32 = undefined;
exports.initializeShaderModules = initializeShaderModules;

var _luma = require('luma.gl');

var _project = require('../shaderlib/project/project');

var _project2 = _interopRequireDefault(_project);

var _project3 = require('../shaderlib/project64/project64');

var _project4 = _interopRequireDefault(_project3);

var _lighting = require('../shaderlib/lighting/lighting');

var _lighting2 = _interopRequireDefault(_lighting);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function initializeShaderModules() {
  (0, _luma.registerShaderModules)([_luma.fp32, _luma.fp64, _project2.default, _project4.default, _lighting2.default, _luma.picking]);

  (0, _luma.setDefaultShaderModules)([_project2.default]);
} // Copyright (c) 2015 - 2017 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

initializeShaderModules();

exports.fp32 = _luma.fp32;
exports.fp64 = _luma.fp64;
exports.picking = _luma.picking;
exports.project = _project2.default;
exports.project64 = _project4.default;
exports.lighting = _lighting2.default;
