// _deck.gl@5.0.1@deck.gl/dist-es6/react/experimental/viewport-layout.js
var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

import React, { createElement, cloneElement } from 'react';
import { flatten } from '../../core/utils/flatten';

var ViewportLayout = function (_React$Component) {
  _inherits(ViewportLayout, _React$Component);

  function ViewportLayout(props) {
    _classCallCheck(this, ViewportLayout);

    return _possibleConstructorReturn(this, (ViewportLayout.__proto__ || Object.getPrototypeOf(ViewportLayout)).call(this, props));
  }

  // Iterate over viewport descriptors and render viewports at specified positions


  _createClass(ViewportLayout, [{
    key: '_renderChildrenUnderViewports',
    value: function _renderChildrenUnderViewports() {
      // Flatten out nested viewports array
      var viewports = flatten(this.props.viewports, { filter: Boolean });

      // Build a viewport id to viewport index
      var viewportMap = {};
      viewports.forEach(function (viewportOrDescriptor) {
        var viewport = viewportOrDescriptor.viewport || viewportOrDescriptor;
        if (viewport.id) {
          viewportMap[viewport.id] = viewport;
        }
      });

      var originalChildren = React.Children.toArray(this.props.children);
      var lastElement = originalChildren.pop();

      var children = originalChildren.map(function (child, i) {
        // If viewportId is provided, match with viewport
        if (child.props.viewportId) {
          var viewport = viewportMap[child.props.viewportId];

          // TODO - this is too react-map-gl specific
          var newProps = Object.assign({}, child.props, {
            visible: viewport.isMapSynched(),
            width: viewport.width,
            height: viewport.height
          }, viewport.getMercatorParams());

          var clone = cloneElement(child, newProps);

          var _style = {
            position: 'absolute',
            left: viewport.x,
            top: viewport.y,
            width: viewport.width,
            height: viewport.height
          };

          child = createElement('div', { key: 'viewport-component-' + i, style: _style }, clone);
        }

        return child;
      });

      var style = { position: 'absolute', left: 0, top: 0 };
      children.push(createElement('div', { key: 'children', style: style }, lastElement));

      return children;
    }
  }, {
    key: 'render',
    value: function render() {
      // Render the background elements (typically react-map-gl instances)
      // using the viewport descriptors
      var children = this._renderChildrenUnderViewports();
      return createElement('div', {}, children);
    }
  }]);

  return ViewportLayout;
}(React.Component);

export default ViewportLayout;
